### ISSUE TITLE
Dynamic sources/boundary conditions?

### ISSUE THREAD

**Issue by jaberkow** [pylbm/pylbm#95]
Is there any way to set up dynamic boundary conditions or sources that be can be adjusted on a per time-step basis, or is that outside the scope/use case of this project?

**Comment by gouarin** [comment]
@jaberkow you already have the possibility to have dynamic source terms (see https://pylbm.readthedocs.io/en/latest/notebooks/08_advection_reaction.html).

It's not possible for boundary conditions yet but it's not so hard. I will try to make that in the next weeks.

**Comment by jaberkow** [comment]
Ah, let me rephrase:  I meant source terms that can be adjusted by the user and/or python functions external to the simulation object.  For instance, in the in the example you linked, could I set the value of alpha externally during each iterate of the while loop to a value produced by an arbitrary python function of  `sol.m[u]`?

**Comment by gouarin** [comment]
There are two answers depending on the complexity of your function

- it's possible to have source terms which depend on `sol.m[u]`. It is useful when you simulate mutli phase flows for example. Here is an example to compute a gradient on `u`

```
 class consm(sp.IndexedBase):
    is_real = True

u = sp.symbols('u', real=True)
u = consm(u)

def source(dx):
    from pylbm.symbolic import ix, iy
    return (u[ix+1, iy] - u[ix-1, iy])/(2*dx)
```

- The other possibility is to split the `onetimestep`. When pylbm makes the code generation, we compile the `onetimestep` but also all the single methods performed into it: `f2m`, `m2f`, `transport`, `relaxation`, ... (see https://github.com/pylbm/pylbm/blob/master/pylbm/algorithm/base.py). So you can call each step of the `onetimestep` algorithm and make between these steps what you want.

**Comment by jaberkow** [comment]
Ok, thank you.  I have a few more questions about this, but would you rather I move this to the gitter channel?

**Comment by gouarin** [comment]
Yes. It will be easier.

**Comment by maniset** [comment]
Hello @gouarin,

Thanks for this great library.

I have the same question. I cannot find a way to implement a new boundary condition at different time-steps. Do you have any suggestions?

I would appreciate your help on this matter.

Sincerely

**Comment by gouarin** [comment]
Hello @maniset 

Thanks for your message.

Could you open a new issue describing more precisely what you try to make? Or come discuss it on gitter.
